pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstance.java

```diff
@@ -52,23 +53,49 @@ public JavaInstance(ContextImpl contextImpl, Object userClassObject) {
         }
     }
 
-    public JavaExecutionResult handleMessage(Record<?> record, Object input) {
+    public CompletableFuture<JavaExecutionResult> handleMessage(Record<?> record, Object input) {
         if (context != null) {
             context.setCurrentMessageContext(record);
         }
+
+        final CompletableFuture<JavaExecutionResult> future = new CompletableFuture<>();
         JavaExecutionResult executionResult = new JavaExecutionResult();
+
         try {
             Object output;
             if (function != null) {
                 output = function.process(input, context);
             } else {
                 output = javaUtilFunction.apply(input);
             }
-            executionResult.setResult(output);
+
+            if (output instanceof CompletableFuture) {
+                // Function is in format: Function<I, CompletableFuture<O>>
+                ((CompletableFuture) output).whenComplete((obj, throwable) -> {
```

Comment:
This is not applying any backpressure on the consumer. If the function doesn't have a mechanism to self-throttle, it will just blow up the memory.

Reply:
shared the same concern here. I think we need to introduce a setting to limit the max outstanding items (results) being processed.

Reply:
Ok. Thanks for the suggestion.